keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -0,0 +1,1200 @@
+and interacting with commands executed on a container in the cluster. An example of
+`kubectl exec` running the `date` command on an `nginx` pod/container is:
+
+```
+$ kubectl exec nginx -- date
+Tue May 16 03:34:04 PM PDT 2023
+```
+
+### Background: API Server and Kubelet `UpgradeAwareProxy`
+
+In order to route the data streamed between the client and the container, both the
+API Server and Kubelet must proxy these data messages. Both the API Server and the
+Kubelet provide this functionality with the `UpgradeAwareProxy`, which is a reverse
+proxy that knows how to deal with the connection upgrade handshake.
+
+### Proposal: `kubectl` WebSocket Executor and Fallback Executor
+
+This enhancement proposes adding a `WebSocketExecutor` to `kubectl`, implementing
+the WebSocket client using a new subprotol version (`v5.channel.k8s.io`). Additionally,
+we propose creating a `FallbackExecutor` to address client/server version skew. The
+`FallbackExecutor`  first attempts to upgrade the connection to WebSockets
+version five (`v5.channel.k8s.io`) with the `WebSocketExecutor`, then falls back
+to the legacy `SPDYExecutor` version four (`v4.channel.k8s.io`), if the upgrade is
+unsuccessful. Note that this mechanism can require two request/response trips instead
+of one. While the fallback mechanism may require an extra request/response if the
+initial upgrade is not successful, we believe this possible extra roundtrip is justified
+for the following reasons:
+
+1. The upgrade handshake is implemented in low-level SPDY and WebSocket libraries,
+and it is not easily exposed by these libraries. If it is even possible to modify
+the upgrade handshake, the added complexity would not be worth the effort.
+2. The streaming is already IO heavy, so another roundtrip will not substantially
+affect the perceived performance.
+3. As releases increment, the probablity of a WebSocket enabled `kubectl` communicating
+with an older non-WebSocket enabled API Server decreases.
+
+### Proposal: `v5.channel.k8s.io` subprotocol version
+
+As previously mentioned, we propose incrementing the subprotocol version from four
+to five: `v5.channel.k8s.io`. This version will indicate to upstream components
+that `kubectl` is interested in using WebSockets between the client and the API Server
+if it is supported.
+
+### Proposal: API Server `StreamTranslatorProxy`
+
+![Stream Translator Proxy](./stream-translator-proxy-2.png)
+
+Currently, the API Server role within client/container streaming is to proxy the
+data stream using the `UpgradeAwareProxy`. This enhancement proposes to modify the
+SPDY data stream between `kubectl` and the API Server by conditionally adding a
+`StreamTranslatorProxy` at the API Server. If the request is for a WebSocket upgrade
+of version `v5.channel.k8s.io`, the `UpgradeAwareProxy` will delegate to the
+`StreamTranslatorProxy`. This translation proxy terminates the WebSocket connection,
```

Comment:
Why `StreamTranslatorProxy` terminates websocket connection. If the websocket connection is terminated between kubectl - API server, how data transfer will be done via these two components?

Reply:
When I say the connection is **terminated** in the `StreamTranslatorProxy`, I do not mean they stop there. I just mean that the **WebSockets** data messages are received there by the WebSockets server. These received data messages are then proxied onto a SPDY client to continue upstream (to Kubelet and then the Container Runtime).